RelationalAI/DefaultAccessTokenHandler.cs

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2022 RelationalAI, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace RelationalAI
+{
+    // This handler caches tokens in ~/.rai/tokens.json. It will attempt to load
+    // a token from the cache file and if it is not found or has expired, it will
+    // delegate to rest.RequestAccessTokenAsync to retrieve a new token and will save it
+    // in the cache file.
+    public class DefaultAccessTokenHandler : IAccessTokenHandler
+    {
+        private static readonly SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1);
+
+        public Rest Rest { get; set; }
+
+        public ILogger Logger { get; set; }
+
+        public async Task<AccessToken> GetAccessTokenAsync(ClientCredentials creds)
+        {
+            var token = ReadAccessToken(creds);
+            if (token != null && !token.IsExpired)
+            {
+                return token;
+            }
+
+            token = await Rest.RequestAccessTokenAsync(creds);
+            if (token != null)
+            {
+                await WriteAccessTokenAsync(creds, token);
+            }
+
+            return token;
+        }
+
+        private string CacheName()
+        {
+            var envHome = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "HOMEPATH" : "HOME";
+            var home = Environment.GetEnvironmentVariable(envHome);
+
+            return Path.Join(home, ".rai", "tokens.json");
+        }
+
+        private AccessToken ReadAccessToken(ClientCredentials creds)
+        {
+            var cache = ReadTokenCache();
+
+            if (cache.ContainsKey(creds.ClientId))
+            {
+                return cache[creds.ClientId];
+            }
+
+            return null;
+        }
+
+        private Dictionary<string, AccessToken> ReadTokenCache()
+        {
+            try
+            {
+                var data = File.ReadAllText(CacheName());
```

Comment:
this class should fallback to in-memory caching if it cannot find the cache folder

Reply:
that's right, this behavior is handled in the caller method `GetAccessTokenAsync()` in the Rest class but it should be part of the handler implementation, moving it here